### PR TITLE
pathutils_nodejs_support

### PR DESCRIPTION
**Pull request by bung87**
compiled passed on my machine, need review not sure am I doing right.

`addNormalizePath` need fix not work with jsgen.

### CODE REVIEW

**Comment by Araq** [comment]
compiler/pathutils is not a standard lib. is it so useful that it should become one?

**Comment by bung87** [comment]
> compiler/pathutils is not a standard lib. is it so useful that it should become one?

I am trying compile nimpretty to js backend, it depends on this module. dont know if it is possible.
I just thought it just string manipulation.

File: compiler/pathutils.nim [hunk 1]
```diff
@@ -22,25 +28,39 @@ type
 proc isEmpty*(x: AnyPath): bool {.inline.} = x.string.len == 0
 
 proc copyFile*(source, dest: AbsoluteFile) =
-  os.copyFile(source.string, dest.string)
+  when defined(nodejs):
+    fs.createReadStream(source.cstring).pipe(fs.createWriteStream(dest.cstring));
+  else:
+    os.copyFile(source.string, dest.string)
```
Comment (timotheecour) [inline compiler/pathutils.nim:34]: but this is the wrong place to branch; instead make `os.copyFile` work with with `nodejs`
likewise with other procs.

That way is much more useful as it'll make clients of os work with nodejs.

**Comment by timotheecour** [comment]
> I am trying compile nimpretty to js backend

I guess that's a fair use case (ideally nim compiler should work on all reasonable backends, including nodejs), but you'll probably face other issues than pathutils; in any case the branching (c vs nodejs) should be done at the right place, ideally at the lowest level place possible (eg the importc proc) unless there's a performance consideration (eg if nodejs has a builtin way to do something that's more efficient than calling the nim way).

in this case, it's probably os.nim for most of those.

right now `getCurrentDir` is the only place in os.nim with branching for nodejs, but you can add more there and that'll be a generally applicable contribution even if nimpretty ends up not working on nodejs

**Comment by bung87** [comment]
I need listen thoughts, make sure this can be continued, as you mentioned some proc can be place to os.nim, I'll try that so , as far I only have "compile nimpretty to js " goal, maybe it will becoming huge for wrap nodejs libs, when it comes I may create separate nodejs modules wrapper like document listed "modules for js backend"

**Comment by timotheecour** [comment]
> I'll try that so

yay. anything is the spirit of https://github.com/nim-lang/Nim/pull/14095 is highly appreciated. Instead of creating  "modules for js backend", it's best to make existing module work (or work better) for js backend (unless something is js-backend specific of course, like jsconsole)

> document listed "modules for js backend"

see tests/js/tstdlib_imports.nim which attempts to list all modules that at least compile with js backend. Of course compile and being fully compatible is a different story.

**Comment by bung87** [comment]
@timotheecour Thanks for your reply! I guess I have new field that I can create PR for :P

**Comment by dom96** [comment]
My goodness, stop adding nodejs support to Nim. If you want it then create a Nim package.

**Comment by timotheecour** [comment]
A nim package can't work for something like that, it requires some support logic in nim.

**Pull request closed without merging**